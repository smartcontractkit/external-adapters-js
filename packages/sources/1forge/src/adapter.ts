import { Builder, Logger, Requester, Validator } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
} from '@chainlink/ea-bootstrap'
import { DEFAULT_WS_API_ENDPOINT, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: DefaultConfig): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true): string => {
    if (!symbol) return ''
    return subscribe ? `subscribe_to|${symbol}` : `unsubscribe_from|${symbol}`
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.quotes.inputParameters,
      {},
      { shouldThrowError: false },
    )
    if (validator.error) return
    if (Array.isArray(validator.validated.data.quote)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.quote} supplied as quote. Only non-array tickers can be used for WS`,
      )
      return
    }
    if (Array.isArray(validator.validated.data.base)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.base} supplied as base. Only non-array tickers can be used for WS`,
      )
      return
    }
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol}/${convert}`
  }
  const parseResponse = (response: string) => {
    const [type, data] = response.split('|')
    return { type, data: JSON.parse(data) }
  }

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message: any) => {
        if (!message.data || message.data.indexOf('update') === -1) return
        const { data } = parseResponse(message.data)
        return getSubscription(data.s)
      },
      isError: () => false,
      filter: (message: any) => message.data.indexOf('update') !== -1,
      toResponse: (message: any) => {
        const { data } = parseResponse(message.data)
        return Requester.success('1', { data: { result: data.p } })
      },
      onConnectChain: [
        {
          payload: `login|${defaultConfig.apiKey}`,
          filter: (message: any) => message.data.indexOf('post_login_success') !== -1,
        },
      ],
    }
  }
}
