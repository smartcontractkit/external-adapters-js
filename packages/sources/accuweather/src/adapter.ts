import { Builder } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
