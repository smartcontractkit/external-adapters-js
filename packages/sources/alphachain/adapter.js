const { Requester, Validator } = require('@chainlink/external-adapter')
const { util } = require('@chainlink/ea-bootstrap')

const customError = (data) => {
  if (data.status !== '200') return true
  return false
}

const customParams = {
  base: ['base', 'from', 'coin'],
  quote: ['quote', 'to', 'market'],
}

const execute = (input, callback) => {
  const validator = new Validator(input, customParams)
  if (validator.error) return callback(validator.error.statusCode, validator.errored)

  const jobRunID = validator.validated.id
  const url = 'https://alpha-chain2.p.rapidapi.com/data-query'
  const host = 'alpha-chain2.p.rapidapi.com'
  const headers = {
    'content-type': 'application/octet-stream',
    'x-rapidapi-host': host,
    'x-rapidapi-key': util.getRandomRequiredEnv('API_KEY'),
    useQueryString: true,
  }
  const base = validator.validated.data.base.toUpperCase()
  const quote = validator.validated.data.quote.toUpperCase()

  const params = {
    from_symbol: base,
    to_symbol: quote,
    chainlink_node: true,
  }

  const config = {
    url,
    params,
    headers,
  }

  Requester.request(config, customError)
    .then((response) => {
      response.data.result = Requester.validateResultNumber(response.data, ['result'])
      callback(response.status, Requester.success(jobRunID, response))
    })
    .catch((error) => callback(500, Requester.errored(jobRunID, error)))
}

module.exports.execute = execute
