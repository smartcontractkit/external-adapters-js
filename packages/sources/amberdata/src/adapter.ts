import { Requester, Validator, Builder } from '@chainlink/ea-bootstrap'
import {
  AdapterRequest,
  Config,
  ExecuteFactory,
  ExecuteWithConfig,
  MakeWSHandler,
  APIEndpoint,
} from '@chainlink/types'
import { DEFAULT_WS_API_ENDPOINT, makeConfig, NAME } from './config'
import * as endpoints from './endpoint'
import { crypto } from './endpoint'

// Export function to integrate with Chainlink node
export const execute: ExecuteWithConfig<Config> = async (request, context, config) => {
  return Builder.buildSelector(request, context, config, endpoints)
}

export const endpointSelector = (request: AdapterRequest): APIEndpoint =>
  Builder.selectEndpoint(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (defaultConfig?: Config): MakeWSHandler => {
  const subscriptions: any = {}
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(input, crypto.inputParameters, {}, false)
    if (validator.error) return
    const base = (validator.overrideSymbol(NAME) as string).toLowerCase()
    const quote = validator.validated.data.quote.toLowerCase()
    return `${base}_${quote}`
  }
  const getSubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'subscribe', params: ['market:tickers', { pair }] }
  }
  const getUnsubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'unsubscribe', params: [subscriptions[pair]] }
  }
  return () => {
    const config = defaultConfig || makeConfig()
    return {
      connection: {
        url: config.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
        protocol: { headers: { ...config.api.headers } },
      },
      subscribe: (input) => getSubscription(getPair(input)),
      unsubscribe: (input) => getUnsubscription(getPair(input)),
      subsFromMessage: (message) => {
        const pair = message?.params?.result?.pair
        subscriptions[pair] = message?.params?.subscription
        return getSubscription(message?.params?.result?.pair)
      },
      // https://github.com/web3data/web3data-js/blob/5b177803cb168dcaed0a8a6e2b2fbd835b82e0f9/src/websocket.js#L43
      isError: () => false, // Amberdata never receives error types?
      filter: (message: any) => !!message.params,
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['params', 'result', 'last'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
