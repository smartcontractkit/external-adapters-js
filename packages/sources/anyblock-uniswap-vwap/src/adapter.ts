import { Execute } from '@chainlink/types'
import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { util } from '@chainlink/ea-bootstrap'

const customError = (data: any) => {
  return !data.hits || !data.hits.hits || data.hits.hits.length < 1
}

const customParams = {
  address: true,
  debug: false,
  roundDay: false,
  start: false,
  end: false,
}

const buildVWAP = (response: any, debug: boolean) => {
  const sources = response.data.hits.hits.map((i: any) => {
    const reserve0 = i._source.args.find((j: any) => j.pos === 0)
    const reserve1 = i._source.args.find((j: any) => j.pos === 1)
    return {
      timestamp: i._source.timestamp,
      reserve0: parseInt(reserve0['value.hex'], 16),
      reserve1: parseInt(reserve1['value.hex'], 16),
    }
  })

  let overallVolume = 0
  let sumAmountAndPrices = 0
  for (let i = 1; i < sources.length; i++) {
    const reserve0volume = Math.abs(sources[i].reserve0 - sources[i - 1].reserve0)
    const price = sources[i].reserve0 / sources[i].reserve1
    overallVolume += reserve0volume
    sumAmountAndPrices += price * reserve0volume
  }

  const vwap = sumAmountAndPrices / overallVolume
  const resp: any = {
    status: response.status,
    data: { result: vwap },
  }

  if (debug) resp.data.raw = sources
  return resp
}

const cleanupDate = (inputDate: string, roundDay: boolean) => {
  let outputDate: number
  try {
    outputDate = parseInt(inputDate)
    if (roundDay) {
      const date = new Date(outputDate)
      date.setUTCHours(0, 0, 0, 0)
      outputDate = date.getTime()
    }
  } catch (err) {
    return inputDate
  }
  return outputDate
}

// TODO: enable other networks
const getAnyblockUrl = () => 'https://api.anyblock.tools/ethereum/ethereum/mainnet/es/event/search/'

export const execute: Execute = async (input) => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  // TODO: validate this is a checksum address
  const address = validator.validated.data.address
  const debug = validator.validated.data.debug || false
  const roundDay = validator.validated.data.roundDay || false
  let start = validator.validated.data.start
  let end = validator.validated.data.end

  const url = getAnyblockUrl()

  end = cleanupDate(end, roundDay)
  start = cleanupDate(start, roundDay)

  if (!start && !end) {
    const date = new Date()
    date.setUTCHours(0, 0, 0, 0)
    end = date.getTime() / 1000

    start = end - 60 * 60 * 24
  } else if (!start) {
    start = end - 60 * 60 * 24
  } else if (!end) {
    end = start + 60 * 60 * 24
  } else if (start === end) {
    start = end - 60 * 60 * 24
  }

  const body = {
    query: {
      bool: {
        filter: [
          { term: { 'address.raw': address } },
          { term: { 'event.raw': 'Sync' } },
          {
            range: {
              timestamp: {
                gte: start,
                lte: end,
              },
            },
          },
        ],
      },
    },
    sort: [{ timestamp: 'asc' }],
    size: 10000,
    _source: ['timestamp', 'args'],
  }

  const headers = {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${util.getRandomRequiredEnv('API_KEY')}`,
  }

  const config = {
    url,
    headers,
    data: body,
  }

  const response = await Requester.request(config, customError)
  const vwapResp = buildVWAP(response, debug)
  return Requester.success(jobRunID, vwapResp)
}
