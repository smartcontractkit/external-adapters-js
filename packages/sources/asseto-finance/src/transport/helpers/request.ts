export interface ApiRequestConfig {
  baseURL: string
  url?: string
  headers?: Record<string, string>
  method?: string
  data?: any
}

export class RequestHelper {
  static createAuthenticatedGetRequest(
    baseURL: string,
    bearerToken: string,
    path?: string,
  ): ApiRequestConfig {
    return {
      baseURL,
      url: path,
      headers: {
        Authorization: `Bearer ${bearerToken}`,
      },
    }
  }

  static createNavRequest(baseURL: string, bearerToken: string, fundId: number): ApiRequestConfig {
    return this.createAuthenticatedGetRequest(baseURL, bearerToken, `/${fundId}/nav-daily`)
  }

  static createReserveRequest(
    baseURL: string,
    bearerToken: string,
    fundId: number,
  ): ApiRequestConfig {
    return this.createAuthenticatedGetRequest(baseURL, bearerToken, `/${fundId}/reserves`)
  }
}
