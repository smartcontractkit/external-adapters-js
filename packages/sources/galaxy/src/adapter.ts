import {
  AdapterConnectionError,
  AdapterDataProviderError,
  AdapterError,
  Builder,
  Logger,
  Requester,
  Validator,
} from '@chainlink/ea-bootstrap'
import type {
  Config,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
  Method,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'
import { AccessTokenResponse, Pair, TickerMessage } from './types'

export const execute: ExecuteWithConfig<Config, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<Config, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<Config, endpoints.TInputParameters> =>
  Builder.selectEndpoint<Config, endpoints.TInputParameters>(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<Config, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const getAccessToken = async (config: Config): Promise<string> => {
  try {
    const tokenResponse = await Requester.request<AccessTokenResponse>({
      url: config.api?.baseURL,
      method: 'get' as Method,
      headers: {
        'X-GALAXY-APIKEY': (config.adapterSpecificParams?.apiKey || '') as string,
        'X-GALAXY-PASSWORD': (config.adapterSpecificParams?.apiPassword || '') as string,
      },
    })
    if (!tokenResponse.data.token)
      throw new AdapterDataProviderError({ message: tokenResponse.data.message || 'Login failed' })
    return tokenResponse.data.token
  } catch (e: any) {
    const err = e as any
    const message = `Login failed ${err.message ? `with message '${err.message}'` : ''}`
    const error = { ...err, message }
    Logger.debug(message)
    throw error.response
      ? new AdapterDataProviderError(error)
      : error.request
      ? new AdapterConnectionError(error)
      : new AdapterError(error)
  }
}

export const makeWSHandler = (defaultConfig?: Config): MakeWSHandler<any> => {
  let token: string | undefined
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.price.inputParameters,
      {},
      { shouldThrowError: false },
    )
    const { base = '', quote = '' } = validator.validated.data
    return base && quote && `markPrice_${base.toUpperCase()}/${quote.toUpperCase()}`
  }

  return async () => {
    const config = defaultConfig || makeConfig()
    if (!token) token = await getAccessToken(config)

    return {
      connection: {
        url: config.ws?.baseWsURL,
        protocol: { headers: { ...config.api?.headers, token } },
      },
      noHttp: true,
      subscribe: (input: AdapterRequest) => ({ type: 'subscribe', signals: [getPair(input)] }),
      unsubscribe: (input: AdapterRequest) => ({ type: 'unsubscribe', signals: [getPair(input)] }),
      subsFromMessage: (message: TickerMessage, subscriptionMessage: Pair) => {
        if (message.type !== 'signal_update') return
        if (!subscriptionMessage.signals.includes(message.signal)) return
        return { type: 'subscribe', signals: subscriptionMessage.signals }
      },
      isError: (message: TickerMessage) => message.type === 'error',
      filter: (message: TickerMessage) => message.type === 'signal_update',
      toResponse: (message: TickerMessage, input: AdapterRequest) => {
        const result = Requester.validateResultNumber(message, ['value'])
        return Requester.success(input.id, { data: { result, ts: message.ts } }, config.verbose)
      },
    } as any // TODO: connection type mismatch
  }
}
