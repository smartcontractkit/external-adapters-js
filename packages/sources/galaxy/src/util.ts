import { makeLogger } from '@chainlink/external-adapter-framework/util'
import {
  AdapterConnectionError,
  AdapterDataProviderError,
  AdapterError,
} from '@chainlink/external-adapter-framework/validation/error'
import axios from 'axios'
import { config } from './config'

const logger = makeLogger('GalaxyUtil')

export interface AccessTokenResponse {
  token?: string
  message?: string
}

export interface AccessToken {
  token: string
  created: number
}

export const getAccessToken = async (settings: typeof config.settings): Promise<AccessToken> => {
  const requestedTs = Date.now()
  try {
    const tokenResponse = await axios.request<AccessTokenResponse>({
      url: settings.API_ENDPOINT,
      method: 'GET',
      headers: {
        'X-GALAXY-APIKEY': settings.WS_API_KEY,
        'X-GALAXY-PASSWORD': settings.WS_API_PASSWORD,
      },
    })

    if (!tokenResponse.data.token)
      throw new AdapterDataProviderError(
        { message: tokenResponse.data.message || 'Login failed' },
        {
          providerDataRequestedUnixMs: requestedTs,
          providerIndicatedTimeUnixMs: undefined,
          providerDataReceivedUnixMs: Date.now(),
        },
      )
    return {
      token: tokenResponse.data.token,
      created: new Date().getTime(),
    }
  } catch (e: unknown) {
    const err = e as any
    const message = `Login failed ${err.message ? `with message '${err.message}'` : ''}`
    const error = { ...err, message }
    logger.debug(message)
    throw error.response
      ? new AdapterDataProviderError(error, {
          providerDataRequestedUnixMs: requestedTs,
          providerIndicatedTimeUnixMs: undefined,
          providerDataReceivedUnixMs: Date.now(),
        })
      : error.request
      ? new AdapterConnectionError(error, {
          providerDataRequestedUnixMs: requestedTs,
          providerIndicatedTimeUnixMs: undefined,
          providerDataReceivedUnixMs: Date.now(),
        })
      : new AdapterError(error)
  }
}
