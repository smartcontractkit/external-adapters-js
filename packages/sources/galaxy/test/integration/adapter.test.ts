import * as process from 'process'
import { AddressInfo } from 'net'
import {
  mockWebSocketProvider,
  mockPriceWebSocketServer,
  createAdapter,
  setEnvVariables,
} from './setup'
import request, { SuperTest, Test } from 'supertest'
import { Server } from 'mock-socket'
import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { AdapterRequestBody, sleep } from '@chainlink/external-adapter-framework/util'
import { WebSocketClassProvider } from '@chainlink/external-adapter-framework/transports'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { mockTokenResponse } from './fixtures'

describe('Price Endpoint', () => {
  let fastify: ServerInstance | undefined
  let req: SuperTest<Test>
  let mockPriceWsServer: Server | undefined
  let spy: jest.SpyInstance
  const tokenEndpoint = process.env.API_ENDPOINT || 'https://test-url.com'
  const wsEndpoint = process.env.WS_API_ENDPOINT || 'ws://localhost:9090'

  jest.setTimeout(10000)

  const priceData: AdapterRequestBody = {
    data: {
      base: 'ETH',
      quote: 'USD',
    },
  }

  let oldEnv: NodeJS.ProcessEnv

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env['WS_SUBSCRIPTION_TTL'] = '5000'
    process.env['CACHE_MAX_AGE'] = '5000'
    process.env['CACHE_POLLING_MAX_RETRIES'] = '0'
    process.env['METRICS_ENABLED'] = 'false'
    process.env['API_ENDPOINT'] = tokenEndpoint
    process.env['WS_API_ENDPOINT'] = wsEndpoint
    process.env['WS_API_KEY'] = 'test-key'
    process.env['WS_API_PASSWORD'] = 'test-password'
    process.env['RATE_LIMIT_CAPACITY_SECOND'] = '2'
    const mockDate = new Date('2022-05-10T16:09:27.193Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())

    mockTokenResponse()
    mockWebSocketProvider(WebSocketClassProvider)
    mockPriceWsServer = mockPriceWebSocketServer(wsEndpoint)

    fastify = await expose(createAdapter() as unknown as Adapter)
    req = request(`http://localhost:${(fastify?.server.address() as AddressInfo).port}`)

    // Send initial request to start background execute
    await req.post('/').send(priceData)
    await sleep(5000)
  })

  afterAll((done) => {
    spy.mockRestore()
    setEnvVariables(oldEnv)
    mockPriceWsServer?.close()
    fastify?.close(done())
  })

  it('should return success', async () => {
    const makeRequest = () =>
      req
        .post('/')
        .send(priceData)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)

    const response = await makeRequest()
    expect(response.body).toEqual({
      result: 1279.2012582120603,
      statusCode: 200,
      data: { result: 1279.2012582120603 },
      timestamps: {
        providerDataReceivedUnixMs: 1652198967193,
        providerDataStreamEstablishedUnixMs: 1652198967193,
        providerIndicatedTimeUnixMs: 1667970828970,
      },
    })
  }, 30000)
  it('should return error (empty body)', async () => {
    const makeRequest = () =>
      req
        .post('/')
        .send({})
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)

    const response = await makeRequest()
    expect(response.statusCode).toEqual(400)
  }, 30000)
  it('should return error (empty data)', async () => {
    const makeRequest = () =>
      req
        .post('/')
        .send({ data: {} })
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)

    const response = await makeRequest()
    expect(response.statusCode).toEqual(400)
  }, 30000)
  it('should return error (empty base)', async () => {
    const makeRequest = () =>
      req
        .post('/')
        .send({ data: { quote: 'BTC' } })
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)

    const response = await makeRequest()
    expect(response.statusCode).toEqual(400)
  }, 30000)
  it('should return error (empty quote)', async () => {
    const makeRequest = () =>
      req
        .post('/')
        .send({ data: { base: 'ETH' } })
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)

    const response = await makeRequest()
    expect(response.statusCode).toEqual(400)
  }, 30000)
})
