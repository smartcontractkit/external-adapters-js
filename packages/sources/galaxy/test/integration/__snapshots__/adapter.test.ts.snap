// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Endpoint should return success 1`] = `
{
  "data": {
    "result": 1279.2012582120603,
  },
  "result": 1279.2012582120603,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": Any<Number>,
    "providerDataStreamEstablishedUnixMs": Any<Number>,
    "providerIndicatedTimeUnixMs": 1667970828970,
  },
}
`;
